service/storage_proxy.cc

```diff
@@ -5254,6 +5262,25 @@ const db::hints::host_filter& storage_proxy::get_hints_host_filter() const {
     return _hints_manager.get_host_filter();
 }
 
+void storage_proxy::mark_responded(const gms::inet_address& ep, clock_type::time_point response_time) {
+    _last_seen[ep].last_responded = response_time;
+}
+
+void storage_proxy::mark_sent(const gms::inet_address& ep, clock_type::time_point req_time) {
+    last_seen_info& info = _last_seen[ep];
+    // If last sending attempt was sufficiently old, the response
+    // info should be reset in order to avoid having stale records.
+    if (info.last_sent + 5s < req_time) {
+        info.last_responded = last_seen_info::unknown;
+    }
```

Comment:
Hm, it's been a month since I wrote all these cases, so I don't remember what I had in mind exactly, but I think I added this check after observing an anomaly when testing manually - e.g. something like that comes to mind:
1. A node gets extremely overloaded and stops responding to coordinators, but it's live enough not to get marked as DOWN
2. A coordinator tries communicating with it for a while and fails. It records that there were some communication attempts at time `T`
3. The coordinator stops communicating with the again, because clients' requests stopped coming
4. (1000 seconds pass)
5. The node regains health and is able to respond in time
6. The coordinator wants to communicate with the replica again - but it sees that it last sent the information over 1000 seconds ago and still hasn't received any response, so it qualifies the node as dead/overloaded, while it's perfectly healthy now

Reply:
The scenarios above are not supposed to be likely because the database of recent timestamps is reset each time a node is reported as added/removed/dead/alive, but I think it may still happen